# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/gfxUserFontSet.cpp
# Commit: 8ed7c109c239
# Full Hash: 8ed7c109c239b07a02e7f3139f075c8f1d3ff0a0
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2019-09-20 21:52:25
# Description:
#   Bug 1580690 - Ensure src:local() entries in the user font set are refreshed if the platform font list is rebuilt. r=jwatt
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D46580
# ==============================================================================

diff -r 932d3794c5d4 -r 8ed7c109c239 gfx/thebes/gfxUserFontSet.cpp
--- a/gfx/thebes/gfxUserFontSet.cpp	Thu Sep 19 17:49:39 2019 +0000
+++ b/gfx/thebes/gfxUserFontSet.cpp	Fri Sep 20 15:24:33 2019 +0000
@@ -1138,6 +1138,24 @@
   return family;
 }
 
+void gfxUserFontSet::ForgetLocalFaces() {
+  for (auto iter = mFontFamilies.Iter(); !iter.Done(); iter.Next()) {
+    const auto fam = iter.Data();
+    const auto& fonts = fam->GetFontList();
+    for (const auto& f : fonts) {
+      auto ufe = static_cast<gfxUserFontEntry*>(f.get());
+      // If the user font entry has loaded an entry using src:local(),
+      // discard it as no longer valid, and reset the load state so that
+      // the load will be re-done based on the updated font list.
+      if (ufe->GetPlatformFontEntry() &&
+          ufe->GetPlatformFontEntry()->IsLocalUserFont()) {
+        ufe->mPlatformFontEntry = nullptr;
+        ufe->LoadCanceled();
+      }
+    }
+  }
+}
+
 ///////////////////////////////////////////////////////////////////////////////
 // gfxUserFontSet::UserFontCache - re-use platform font entries for user fonts
 // across pages/fontsets rather than instantiating new platform fonts.